ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/dataflow/reader_mcast_receiver_unary_sharded_ln_pre_allgather.cpp

```diff
@@ -0,0 +1,191 @@
+                if constexpr(row_major) {
+                    ++x;
+                    if (x == num_x) {
+                        x = 0;
+                        ++y;
+                        if (y == num_y) {
+                            y = 0;
+                        }
+                    }
+                } else {
+                    ++y;
+                    if (y == num_y) {
+                        y = 0;
+                        ++x;
+                        if (x == num_x) {
+                            x = 0;
+                        }
+                    }
+                }
+            }
+        }
+    } else {
+        remote_noc_addrs_first_stage[0] = get_noc_addr(in0_remote_noc_x[0], in0_remote_noc_y[0], 0);
+        remote_noc_addrs_second_stage[0] = get_noc_addr(in0_remote_noc_x[0], in0_remote_noc_y[0], 0);
+    }
+
+    volatile tt_l1_ptr uint32_t* reduce_receiver_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(reduce_receiver_semaphore_addr);
+    volatile tt_l1_ptr uint32_t* reduce_sender_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(reduce_sender_semaphore_addr);
+    volatile tt_l1_ptr uint32_t* reduce_second_stage_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(reduce_second_stage_semaphore_addr);
+
+    const uint64_t reduce_receiver_semaphore_noc_addr = get_noc_addr(in0_remote_noc_x[0], in0_remote_noc_y[0], reduce_receiver_semaphore_addr);
+    const uint64_t reduce_second_stage_receiver_semaphore_noc_addr = remote_noc_addrs_second_stage[0] | reduce_second_stage_semaphore_addr;
+
+    const auto& global_reduce_receiver = [&](const uint32_t cb_partial, const uint32_t cb_external, const uint32_t cb_reduce_first_stage) __attribute__((always_inline))
+    {
+        uint32_t num_tiles_per_partial_result = 2;
+        #ifdef RMSNORM
+        num_tiles_per_partial_result = 1;
+        #endif
+        // global reduce
+        // wait for local data ready
+        cb_wait_front(cb_partial, num_tiles_per_partial_result*block_h); // two tiles * block_h
+
+
+        // inc mcast sender
+        noc_semaphore_set(reduce_sender_semaphore_addr_ptr, INVALID);
+        noc_semaphore_inc(reduce_receiver_semaphore_noc_addr, 1);
+        noc_semaphore_wait(reduce_sender_semaphore_addr_ptr, VALID);
+
+
+        if constexpr (is_all_to_all_worker) {
+            // read data from other cores - reduce first stage
+            uint32_t l1_read_addr_ex_par = get_read_ptr(cb_partial);
+            l1_read_addr_ex_par += all_to_all_tile_offset_bytes;
+            // read data from other cores - second stage reduce
+            uint32_t l1_read_addr_ex = 0;
+            uint32_t block_index_stride = 0;
+            if constexpr(use_two_stage_reduce) {
+                l1_read_addr_ex = get_read_ptr(cb_reduce_first_stage);
+                if constexpr(row_major) {
+                    block_index_stride = num_x;
+                } else {
```

Comment:
Considering Full Activation of 32x8k on 4 distributed devices.
Activation shape on each device (32, 2k).
If you shard (32, 2k) on a core grid of (2, 8).
There will be two  `is_all_to_all_worker`
It will reach this code section for core coordinate (1, 0).

Reply:
yeah that's right